docs/reference/scripting/using.asciidoc

```diff
@@ -1,425 +1,471 @@
+real magic is that a `def` type variable can change the type it represents
+during the compilation and evaluation of a script. This capability is incredibly
+powerful when your script is retrieving data, especially `_doc` values.
+
+We'll talk more about retrieving data later on, but the main takeaway relating
+to `def` values is this: you don't really need to know what primitive type your
+data is; you can just call a `def` type and then start operating on the data.
+
+For example, the following request starts with a list that includes primitive
+types of `int`, `double`, and `String`. The request then declares an empty
+`String` and introduces a `for` loop. This loop iterates through the values in
+the list and changes the value three times:
+
+* `1` is an `int`
+* `2.0` is a `double`
+* `'3'` is a `string`
 
 [source,console]
 ----
-POST /_reindex
+POST /_scripts/painless/_execute
 {
-  "source": {
-    "index": "my_test_scores"
-  },
-  "dest": {
-    "index": "my_test_scores_2",
-    "pipeline": "my_test_scores_pipeline"
+  "script": {
+    "source": """
+      List list = [1, 2.0, '3'];
+      String string = "";
+      for (def value : list) {
+        string += value.toString();
+        }
+    """
   }
 }
 ----
-// TEST[continued]
 
-Continue using the pipeline to index any new documents to `my_test_scores_2`.
+The type changes occur automatically without doing anything other than using
+`def` and iterating through the values. Painless changes the type for you so
+that you can focus on writing your script without worrying about what your data
+type is.
 
-[source,console]
-----
-POST /my_test_scores_2/_doc/?pipeline=my_test_scores_pipeline
-{
-  "student": "kimchy",
-  "grad_year": "2099",
-  "math_score": 800,
-  "verbal_score": 800
-}
-----
-// TEST[continued]
+[discrete]
+[[types-whats-next]]
+==== What's next
+These types are the most basic structures in Painless. There are {painless}/painless-types.html[other types] that you can learn about as you script writing
+skills develop. Now that you've learned about types and variables, you can
+start combining them with operators.
 
-These changes may slow indexing but allow for faster searches. Users can now
-sort searches made on `my_test_scores_2` using the `total_score` field instead
-of using a script.
+<<script-spec-operators,Learn about operators>>
 
-[source,console]
-----
-GET /my_test_scores_2/_search
-{
-  "query": {
-    "term": {
-      "grad_year": "2099"
-    }
-  },
-  "sort": [
-    {
-      "total_score": {
-        "order": "desc"
-      }
-    }
-  ]
-}
-----
-// TEST[continued]
+[[script-spec-operators]]
+=== Operators
+An operator is the most basic action to evaluate variables and values in a
+script. Painless offers basic operators for numeric computations
+such as addition, division, and subtraction, but also provides operators for
+comparisons, conditionals, and {painless}/painless-operators.html[much more].
```

Comment:
String concatenation is important, should we mention it here?

Reply:
As discussed, removing the Operators section.